pkg/rc/replication_controller.go

```diff
@@ -316,6 +308,72 @@ func (rc *replicationController) addPods(current types.PodLocations, eligible []
 	return nil
 }
 
+func (rc *replicationController) doNotEnoughNodes(currentNodes []types.NodeName, eligibleNodes []types.NodeName, numScheduledNodes int, txn *auditingTransaction) error {
+	var err error = nil
+	if rc.AllocationStrategy == fields.CattleStrategy {
+		err = rc.doNotEnoughCattleNodes(currentNodes, eligibleNodes, numScheduledNodes)
+	} else {
+		err = rc.doNotEnoughPetNodes(currentNodes, eligibleNodes, numScheduledNodes)
+	}
+
+	// commit any queued operations
+	ok, resp, txnErr := txn.Commit(rc.txner)
+	switch {
+	case txnErr != nil:
+		return txnErr
+	case !ok:
+		return util.Errorf("could not schedule pods due to transaction violation: %s", transaction.TxnErrorsToString(resp.Errors))
+	}
+
+	return err
+}
+
+func (rc *replicationController) doNotEnoughCattleNodes(currentNodes []types.NodeName, eligibleNodes []types.NodeName, numScheduledNodes int) error {
```

Comment:
I don't see any code to check that the RC lock is still held. Is this code run inside of the RC lock's critical section at all?

Reply:
Ah I see that the lock is used when accessing the fields in the `fields.RC` struct. I'll update to use the lock.